Blog post about building ROOT with distcc (#383)
root-project/web: blog/_posts/2020-08-12-building-root-with-distcc.md

---
title: "Building ROOT fast - distributed compilation with distcc"
layout: archive
author: Stefan Wunsch
---

> Hello there, here is Stefan with a short guide on how you can compile ROOT distributed with distcc!

Reviewer: ```suggestion
Hello there, here is Stefan with a short guide on how you can distribute ROOT compilation with distcc!
```

> ROOT is a huge project with more than 24 thousand files and above 4.6 million lines of code (counted with [`cloc`](https://github.com/AlDanial/cloc)). Consequently, a full rebuild from scratch takes a while! To give some numbers:

Reviewer: I guess this is counting LLVM as part of ROOT? That is a bit misleading w.r.t. the actual size of ROOT.

I get:

```
~/W/R/root (print_jitted_code) fd --extension cxx --extension h --extension C --exclude interpreter | wc -l                                                                   
7812                                                                                                                                                                          
~/W/R/root (print_jitted_code) fd --extension cxx --extension h --extension C  | wc -l                                                                                        
10525                                                                                   
```

Similarly I get 1.7M SLOC with `cloc --force-lang cpp (fd --exclude interpreter --extension cxx --extension h --extension C)` (I might be doing something wrong/different, but 4.6M seems to be an overestimation?).

Author: It's with LLVM, yep! Well, I don't know what to write here. At the end we compile almost 5 MLOC, so it's not wrong ;) Even though ROOT itself has probably only 2 MLOC, as you showed. I agree, a little bit misleading, but I'm not sure how to write it differently/better.

Reviewer: Even with LLVM I get 10k source files, not 24k.

Something like this maybe:

> ROOT itself consists of about 1.7M lines of code in almost 8k source files. On top of that, our C++ interpreter is based on clang and LLVM, so that is another large project that is compiled as a dependency when building ROOT.

Author: ```
[stefan@archi root-dev]$ cloc . --exclude-dir llvm
   17090 text files.
   16400 unique files.                                          
    3300 files ignored.

github.com/AlDanial/cloc v 1.86  T=14.71 s (988.7 files/s, 251009.1 lines/s)
---------------------------------------------------------------------------------------
Language                             files          blank        comment           code
---------------------------------------------------------------------------------------
C++                                   5809         299361         363421        1327958
C/C++ Header                          5472         127392         166787         535624
C                                      302          40062          41787         240538
XML                                    800           5435           1622         146246
JavaScript                              77          15350           5517          55943
Python                                 496          13614          16067          49145
Markdown                               222          13981              1          46897
CMake                                  501           4152           5601          25610
HTML                                   197           2329            194          20372
Bourne Shell                           101           1868           1941          17176
Objective C++                           26           3572           3088          13250
Fortran 77                               4            351            540          12570
make                                   284           4847           3798          11416
TeX                                     82           1403           1550          10552
Windows Module Definition                9            181              4           6159
CSS                                     21            566            197           4907
CUDA                                    22           1016           1712           4306
reStructuredText                        33            980            359           3041
m4                                      16            331            506           2355
Objective C                              5            157             89            834
C Shell                                  5             98            136            489
YAML                                     8             68             63            441
Jupyter Notebook                        12              0           2344            370
awk                                      1              1             94            355
JSON                                    10              2              0            288
DOS Batch                                2             35              5            240
Windows Resource File                    4              9              5            173
diff                                     7             31            156            161
Bourne Again Shell                       4             47             28            141
DTD                                      3             25             38            135
SVG                                      1              1              1             80
Fish Shell                               1             10              9             40
NAnt script                              1              1              0             14
SQL                                      1              0              0             14
SWIG                                     2              4              0             14
R                                        1              2              3              8
Perl                                     1              3              4              7
XSLT                                     1              2              7              7
D                                        2              2              0              6
---------------------------------------------------------------------------------------
SUM:                                 14546         537289         617674        2537882
---------------------------------------------------------------------------------------
```

Author: i think 14k files and 2.5 MLOC is fair!

> ROOT is a huge project with more than 24 thousand files and above 4.6 million lines of code (counted with [`cloc`](https://github.com/AlDanial/cloc)). Consequently, a full rebuild from scratch takes a while! To give some numbers:

Reviewer: ```suggestion
ROOT is a huge project with more than 24 thousand files and above 4.6 million lines of code (counted with [`cloc`](https://github.com/AlDanial/cloc)). Consequently, a full rebuild from scratch takes a while! To give some numbers, here is what happens on my laptop:
```

> ROOT is a huge project with more than 24 thousand files and above 4.6 million lines of code (counted with [`cloc`](https://github.com/AlDanial/cloc)). Consequently, a full rebuild from scratch takes a while! To give some numbers:

Reviewer 2: ```suggestion
ROOT is a huge project with more than 24 thousand files and 4.6 million lines of code (counted with [`cloc`](https://github.com/AlDanial/cloc)). That's because it comes with a full compiler, which [interprets C++ code](https://root.cern/primer/#learn-c-at-the-root-prompt). Consequently, a full rebuild from scratch takes a while! To give some numbers:
```

Author: Already took the numbers without LLVM!

```
Building ROOT on 4 cores: 86 minutes
```

> How to make this faster? Besides buying a more powerful computer, another answer from the software community is [`distcc`](https://distcc.github.io/). `distcc` allows to compile C, C++ and objective C(++) distributed, is simple to set up and already integrated into the ROOT build system!

Reviewer: I think this is a good point to clarify an important pre-requisite: you need another machine ("distributed compilation" might not be familiar to some readers, but this should clarify what you are going to talk about).
Feel free to rephrase differently than how I suggest below, of course.

```suggestion
How to make this faster? Besides buying a more powerful computer, another answer from the software community is [`distcc`](https://distcc.github.io/). If you have a spare computer or a second computer reachable over the network, `distcc` allows to distribute compilation of C, C++ and objective C(++) code over multiple machines. It is simple to set up and has been recently integrated into the ROOT build system!
```

Author: I thought `distributed` tells already that there is more than one dedicated machine ;) But most likely you have a point! Never hurts to be more specific :)

## Setup

> But first we have to make another machine accessible for the distributed compilation, called a *volunteer* in the `distcc` terminology. Note that `distcc` distributes only the compilation step, not the preprocessing (without the [`pump` mode](https://distcc.github.io/faq.html)) and linking. This means that the volunteer does not need any dependencies installed, making the setup simple!

Reviewer: ```suggestion
But first we have to make another machine accessible for the distributed compilation, called a *volunteer* in the `distcc` terminology. Note that `distcc` distributes only the compilation step, not the preprocessing (without the [`pump` mode](https://distcc.github.io/faq.html)) and linking. This means that the volunteer does not need any dependencies installed nor the machines have to share a filesystem, making the setup simple!
```

Author: Very good point!

> But first we have to make another machine accessible for the distributed compilation, called a *volunteer* in the `distcc` terminology. Note that `distcc` distributes only the compilation step, not the preprocessing (without the [`pump` mode](https://distcc.github.io/faq.html)) and linking. This means that the volunteer does not need any dependencies installed, making the setup simple!

Reviewer 2: ```suggestion
But first we have to make another machine accessible for the distributed compilation, called a *volunteer* in the `distcc` terminology. Note that `distcc` distributes only the compilation step, not the preprocessing (unless you use the [`pump` mode](https://distcc.github.io/faq.html)) and linking. This means that the volunteer does not need any dependencies installed, making the setup simple!
```

```bash
# Set up distcc on a second machine (a volunteer) using gcc as compiler
# Take care that gcc has the same version than on the main system!
sudo apt install gcc distcc # Or use your preferred package manager!

# Run the distcc(d) daemon!
distccd --daemon --allow-private
```

Now, let's set up your environment on the main system to detect the volunteer!

```bash
# Tell distcc where to find other servers and how many cores from the volunteers
# should be used (for me it's the ip 192.168.1.225 and 12 cores)
# Note that distcc communicates by default via the port 3632, which may be blocked
# by your firewall!
export DISTCC_HOSTS="localhost/4 192.168.1.225/12"

# Clone ROOT (assuming all dependencies are installed)

> git clone https://www.github.com/root-project/root

Reviewer: I think the canonical address is

```suggestion
git clone https://github.com/root-project/root.git
```

cd root/

# Configure ROOT to use distcc
# You can also use distcc together with `-Dccache=ON` to enable caching, which is
# checked before the compilations are distributed.
cmake . -DCMAKE_CXX_COMPILER="g++" -DCMAKE_C_COMPILER="gcc" -Ddistcc="ON"

# Run the compilation distributed on 4+12 cores!
make -j16
```

## Docker

> Most likely you want to make use of an existing system as a (part time) volunteer, which means that installing exactly the same compiler than on the main system can be non trivial. Luckily, [Docker](https://www.docker.com/) makes deploying services such as `distccd` simple and you can even use a Windows machine as volunteer! Just create the following [`Dockerfile`](https://docs.docker.com/engine/reference/builder/) and follow the instructions below.

Reviewer: > installing exactly the same compiler than on the main system can be non trivial

The setup section should probably mention that this is a requirement

Author: It does! I think it's in the bash part.

Reviewer: ah you are right, I had missed it

```dockerfile
FROM ubuntu:focal
RUN apt update -y && apt install -y gcc distcc htop && useradd distcc
ENTRYPOINT distccd --allow-private --daemon && htop

> ```

Reviewer: might be worth specifying that this is if your first machine is ubuntu focal (I think?)

Author: well, the idea is that you just pick a system that has the compiler you need. it does not have to match your distro/OS. but at the end you'll pick the same one than you run because it's the trivial solution to the problem!

Reviewer: sure, still I think it would be good to specify that this is not a dockerfile that will work for everyone: you need to pick a `FROM` with the same (or a similar enough) platform as your main machine.

```bash
# Build the image with the dockerfile above (using Ubuntu 20.04 as the base image)
docker build -t distcc .

> # Run the container with distccd and show htop

Reviewer: ```suggestion
# On the volunteer, run the container with distccd and show htop
```

docker run --rm -ti -p 3632:3632 distcc
```

## Monitoring

`distcc` typically comes with the monitoring application `distccmon`, either text based (`distccmon-text`) or with a GUI (`distccmon-gnome`). Here is a screenshot of the monitoring during my build process showing the compilation of the inbuilt LLVM:

![distcc monitor]({{ '/assets/images/building_root_with_distcc.png' | relative_url }})

> ## Is it now faster?

Reviewer 2: ```suggestion
## Is it faster now?
```

On my setup I get a speedup of 3.2! The compile time goes down from 1 hour and 26 minutes to just 27 minutes.

```
Building ROOT on 4+12 cores: 27 minutes
Speedup: 86 / 27 = 3.2 times faster!
```

> The setup achieves approximately 80% of the optimal scaling of 4, which does not take into account that preprocessing and linking happens only on the main system. Probably `distcc`'s [`pump` mode](https://distcc.github.io/faq.html) would improve the scaling even further? Feel free to try and tell us in the [ROOT forum](https://root-forum.cern.ch/)!

Reviewer 2: ```suggestion
With this setup, I achieved 80% of the theoretical speed up of 4, which does not take into account that preprocessing and linking happens only on the main system. `distcc`'s [`pump` mode](https://distcc.github.io/faq.html) may improve the scaling even further. Feel free to try and tell us in the [ROOT forum](https://root-forum.cern.ch/)!
```
